// unleash-client/lib/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get = exports.post = exports.buildHeaders = exports.getDefaultAgent = void 0;
var fetch = require("make-fetch-happen");
var http = require("http");
var https = require("https");
var httpAgent = new http.Agent({
    keepAlive: true,
    keepAliveMsecs: 30 * 1000,
    timeout: 10 * 1000,
});
var httpsAgent = new https.Agent({
    keepAlive: true,
    keepAliveMsecs: 30 * 1000,
    timeout: 10 * 1000,
});
var getDefaultAgent = function (url) { return (url.protocol === 'https:' ? httpsAgent : httpAgent); };
exports.getDefaultAgent = getDefaultAgent;
var buildHeaders = function (appName, instanceId, etag, contentType, custom, specVersionSupported) {
    var head = {};
    if (appName) {
        head['UNLEASH-APPNAME'] = appName;
        head['User-Agent'] = appName;
    }
    if (instanceId) {
        head['UNLEASH-INSTANCEID'] = instanceId;
    }
    if (etag) {
        head['If-None-Match'] = etag;
    }
    if (contentType) {
        head['Content-Type'] = contentType;
    }
    if (specVersionSupported) {
        head['Unleash-Client-Spec'] = specVersionSupported;
    }
    if (custom) {
        Object.assign(head, custom);
    }
    return head;
};
exports.buildHeaders = buildHeaders;
var post = function (_a) {
    var url = _a.url, appName = _a.appName, timeout = _a.timeout, instanceId = _a.instanceId, headers = _a.headers, json = _a.json, httpOptions = _a.httpOptions;
    return fetch(url, {
        timeout: timeout || 10000,
        method: 'POST',
        agent: (httpOptions === null || httpOptions === void 0 ? void 0 : httpOptions.agent) || exports.getDefaultAgent,
        headers: (0, exports.buildHeaders)(appName, instanceId, undefined, 'application/json', headers),
        body: JSON.stringify(json),
        strictSSL: httpOptions === null || httpOptions === void 0 ? void 0 : httpOptions.rejectUnauthorized,
    });
};
exports.post = post;
var get = function (_a) {
    var url = _a.url, etag = _a.etag, appName = _a.appName, timeout = _a.timeout, instanceId = _a.instanceId, headers = _a.headers, httpOptions = _a.httpOptions, supportedSpecVersion = _a.supportedSpecVersion;
    return fetch(url, {
        method: 'GET',
        timeout: timeout || 10000,
        agent: (httpOptions === null || httpOptions === void 0 ? void 0 : httpOptions.agent) || exports.getDefaultAgent,
        headers: (0, exports.buildHeaders)(appName, instanceId, etag, undefined, headers, supportedSpecVersion),
        retry: {
            retries: 2,
            maxTimeout: timeout || 10000,
        },
        strictSSL: httpOptions === null || httpOptions === void 0 ? void 0 : httpOptions.rejectUnauthorized,
    });
};
exports.get = get;
