// @unleash/proxy/dist/logger.js
"use strict";
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-explicit-any */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleLogger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel["debug"] = "debug";
    LogLevel["info"] = "info";
    LogLevel["warn"] = "warn";
    LogLevel["error"] = "error";
    LogLevel["fatal"] = "fatal";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
const weight = new Map([
    [LogLevel.debug, 0],
    [LogLevel.info, 1],
    [LogLevel.warn, 2],
    [LogLevel.error, 3],
    [LogLevel.fatal, 4],
]);
const resolve = (logLevel) => {
    const w = weight.get(logLevel);
    return w || -1;
};
const stripEmptyArray = (arr) => {
    if (!arr || arr.length === 0) {
        return '';
    }
    return arr;
};
class SimpleLogger {
    constructor(logLevel = LogLevel.warn) {
        this.logLevel = logLevel;
    }
    shouldLog(desired) {
        return resolve(desired) >= resolve(this.logLevel);
    }
    debug(message, ...args) {
        if (this.shouldLog(LogLevel.debug)) {
            console.log(`DEBUG: ${message}`, stripEmptyArray(args));
        }
    }
    info(message, ...args) {
        if (this.shouldLog(LogLevel.info)) {
            console.log(`INFO: ${message}`, stripEmptyArray(args));
        }
    }
    warn(message, ...args) {
        if (this.shouldLog(LogLevel.warn)) {
            console.log(`WARN: ${message}`, stripEmptyArray(args));
        }
    }
    error(message, ...args) {
        if (this.shouldLog(LogLevel.error)) {
            console.log(`ERROR: ${message}`, stripEmptyArray(args));
        }
    }
    fatal(message, ...args) {
        if (this.shouldLog(LogLevel.fatal)) {
            console.log(`FATAL: ${message}`, stripEmptyArray(args));
        }
    }
}
exports.SimpleLogger = SimpleLogger;
