// @unleash/proxy/dist/test/client.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = __importDefault(require("../client"));
const config_1 = require("../config");
const logger_1 = require("../logger");
const unleash_mock_1 = __importDefault(require("./unleash.mock"));
test('should add environment to isEnabled calls', () => {
    let unleashSDK;
    const init = (opts) => {
        unleashSDK = new unleash_mock_1.default(opts);
        return unleashSDK;
    };
    const config = (0, config_1.createProxyConfig)({
        unleashApiToken: '123',
        unleashUrl: 'http://localhost:4242/api',
        proxySecrets: ['s1'],
        environment: 'test',
        logLevel: logger_1.LogLevel.error,
    });
    config.disableMetrics = true;
    const client = new client_1.default(config, init);
    const fakeUnleash = client.unleash;
    fakeUnleash.toggleDefinitions.push({
        name: 'test',
        enabled: false,
        type: 'experiment',
        stale: false,
        strategies: [],
        variants: [],
        impressionData: true,
    });
    client.getEnabledToggles({});
    expect(fakeUnleash.contexts[0].environment).toBe('test');
    client.destroy();
});
test('should override environment to isEnabled calls', () => {
    let unleashSDK;
    const init = (opts) => {
        unleashSDK = new unleash_mock_1.default(opts);
        return unleashSDK;
    };
    const config = (0, config_1.createProxyConfig)({
        unleashApiToken: '123',
        unleashUrl: 'http://localhost:4242/api',
        proxySecrets: ['s1'],
        environment: 'never-change-me',
        logLevel: logger_1.LogLevel.error,
    });
    config.disableMetrics = true;
    const client = new client_1.default(config, init);
    const fakeUnleash = client.unleash;
    fakeUnleash.toggleDefinitions.push({
        name: 'test',
        enabled: false,
        type: 'experiment',
        stale: false,
        strategies: [],
        variants: [],
        impressionData: true,
    });
    client.getEnabledToggles({ environment: 'some' });
    expect(fakeUnleash.contexts[0].environment).toBe('never-change-me');
    client.destroy();
});
